Library/Homebrew/metaformula.rb

```diff
@@ -0,0 +1,56 @@
+require "formula"
+
+# A meta-formula is a special kind of formula which will
+# * Allow bulk management of several related formulae (only `install `now)
+# * extend the semantics of 'dependency' to casks(TODO)
+#
+# As a result, {MetaFormula} differs from {Formula} in that
+# * `url` is not mandatory
+# * `version` is mandatory, (since it can not be deduced from `url` now)
+# * the keg directory can be empty now (which will influence the "empty installation" check in audit.rb)
+
+class MetaFormula < Formula
+  # Allow specs with no url
+  def spec_eval(name)
+    super
+
+    return unless name.equal? :stable
+    # we allow :stable spec with no url here
+    # unnecessary to keep three specs
+    # if we do, {Formula#installed_prefix} will report an error when `brew install xxx`
+    @stable = self.class.send(:stable)
+    @stable.owner = self
+    # `Formula#determine_active_spec` should now raise no "formulae require at least a URL" error
+  end
+
+  # Ignore url error in attribute validation
+  def validate_attributes!
+    super
+  rescue FormulaValidationError => e
+    raise unless e.attr.equal? :url
+  end
+
+  # Metaformula is not designed to be bottled for now
+  def pour_bottle?
+    false
+  end
+
+  # DSL not designed to be used in MetaFormula for now are disabled to avoid undefined behaviors
+  # This is subject to change if certain functions should be assigned to them
+  def self.disable_dsl(name)
+    old_method = method(name)
+    define_singleton_method(name) do |*args, &block|
+      if block
+        opoo "#{name} DSL is ignored in MetaFormula"
+      else
+        old_method.call(*args)
```

Comment:
Why is this passing through the value to the old method?

Reply:
Because some DSL functions (like `Formula::stable`) are also called by other functions (like `Formula::url`).  I think we should only block callings when used as DSL,

Reply:
Sure but as-is this will mean unfortunately that `head "http://..../"`, `bottle :disable` and `patch "http://.../"` in a formula will still work.

Reply:
Oh. That's true. Then how about using caller to detect who calls the function ?
```ruby
if come_from_formula_file(caller(0)[1]) then
   block_it
else
   pass_through
end
```

Reply:
Feels like a bit of a mess/hack. Probably better to define each of these methods separately.

Reply:
It seem we still need to tell the caller even if defining these methods separately. So now I block all calls with arg or block. Do you think it is ok?

Reply:
@gigaflw I still think handling each individual DSL method by itself would be cleaner; I don't think a one-size fits all makes sense given that e.g. `head`, `bottle` and `patch` take single arguments (as above) but `stable` does not.